dbms/src/Flash/Mpp/MPPTaskManager.cpp

```diff
@@ -89,31 +97,22 @@ void MPPTaskManager::cancelMPPQuery(UInt64 query_id, const String & reason)
     LOG_WARNING(log, fmt::format("Begin cancel query: {}", query_id));
     FmtBuffer fmt_buf;
     fmt_buf.fmtAppend("Remaining task in query {} are: ", query_id);
-    // TODO: cancel tasks in order rather than issuing so many threads to cancel tasks
-    std::vector<std::thread> cancel_workers;
-    for (const auto & task : task_set->task_map)
+    auto thread_manager = newThreadManager();
+    for (auto it = task_set->task_map.begin(); it != task_set->task_map.end();)
     {
-        fmt_buf.fmtAppend("{} ", task.first.toString());
-        std::thread t(&MPPTask::cancel, task.second, std::ref(reason));
-        cancel_workers.push_back(std::move(t));
+        fmt_buf.fmtAppend("{} ", it->first.toString());
+        auto current_task = it->second;
+        it = task_set->task_map.erase(it);
```

Comment:
Looks like `isTaskToBeCancelled` needs to be modified too
```
bool MPPTaskManager::isTaskToBeCancelled(const MPPTaskId & task_id)
{
    std::unique_lock lock(mu);
    auto it = mpp_query_map.find(task_id.start_ts);
    if (it != mpp_query_map.end() && it->second->to_be_cancelled)
    {
        return it->second->task_map.find(task_id) != it->second->task_map.end();
    }
    return false;
}
```

Reply:
Done